packages/jest-util/src/TestPathPatterns.ts

```diff
@@ -0,0 +1,110 @@
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import type {Config} from '@jest/types';
+import {escapePathForRegex, replacePathSepForRegex} from 'jest-regex-util';
+
+type PatternsConfig = Pick<Config.GlobalConfig, 'rootDir'>;
+type PatternsFullConfig = PatternsConfig &
+  Pick<Config.GlobalConfig, 'testPathPatterns'>;
+
+export default class TestPathPatterns {
+  readonly patterns: Array<string>;
+  private readonly rootDir: string;
+
+  private _regexString: string | null = null;
+
+  constructor(patterns: Array<string>, config: PatternsConfig);
+  constructor(config: PatternsFullConfig);
+  constructor(
+    patternsOrConfig: Array<string> | PatternsFullConfig,
+    configArg?: PatternsConfig,
+  ) {
+    let patterns, config;
+    if (Array.isArray(patternsOrConfig)) {
+      patterns = patternsOrConfig;
+      config = configArg!;
+    } else {
+      patterns = patternsOrConfig.testPathPatterns;
+      config = patternsOrConfig;
+    }
+
+    this.patterns = patterns;
+    this.rootDir = config.rootDir.replace(/\/*$/, '/');
+  }
+
+  private get regexString(): string {
+    if (this._regexString !== null) {
+      return this._regexString;
+    }
+    const rootDirRegex = escapePathForRegex(this.rootDir);
+    const regexString = this.patterns
+      .map(p => {
+        // absolute paths passed on command line should stay same
+        if (p.match(/^\//)) {
+          return p;
+        }
+
+        // explicit relative paths should resolve against rootDir
+        if (p.match(/^\.\//)) {
+          return p.replace(/^\.\//, rootDirRegex);
+        }
+
+        // all other patterns should only match the relative part of the test
+        return `${rootDirRegex}(.*)?${p}`;
+      })
+      .map(replacePathSepForRegex)
+      .join('|');
+    this._regexString = regexString;
+    return regexString;
+  }
+
+  private get regex(): RegExp {
+    return new RegExp(this.regexString, 'i');
+  }
+
+  /**
+   * Return true if there are any patterns.
+   */
+  isSet(): boolean {
+    return this.patterns.length > 0;
+  }
+
+  /**
+   * Return true if the patterns form a valid regex.
+   */
+  isValid(): boolean {
+    try {
+      // @ts-expect-error noUnusedLocals
+      // eslint-disable-next-line @typescript-eslint/no-unused-vars
+      const _ = this.regex;
+      return true;
+    } catch {
+      return false;
+    }
+  }
+
+  /**
+   * Return true if the given ABSOLUTE path matches the patterns.
+   *
+   * Throws an error if the patterns form an invalid regex (see `isValid`).
+   */
+  isMatch(path: string): boolean {
+    return this.regex.test(path);
+  }
+
+  /**
+   * Return a human-friendly version of the pattern regex.
+   *
+   * Does no normalization or anything, just a naive joining of the regex,
+   * for simplicity.
+   */
+  toPretty(): string {
+    const regex = this.patterns.map(p => p.replace(/\//g, '\\/')).join('|');
```

Comment:
why do we want the escape characters? for copy paste?

I'm not sure I find it very "pretty" 😅

Reply:
ah I was trying to imitate what the original code was doing. Instead of doing this manually, I really should just return `this.regex.toString()`.

And, sure, maybe it's not "pretty" anymore, so I'll rename the function to `toString` (really it's just rendering the patterns back to a string to show to the user)

Reply:
Ah that's right, I can't do `this.regex.toString()` because the regex now contains extra information about the root directory:

```
# Original message
Ran all test suites matching /a\/b|c\/d/i.

# TestPathPatterns.toPretty()
Ran all test suites matching /a\/b|c\/d/i.

# TestPathPatterns.regex.toString()
Ran all test suites matching /\/(.*)?a\/b|\/(.*)?c\/d/i.
```

It's "pretty" in the sense that it's not literally the regex that we're using to match tests. If your main concern is the name of the function, I'm happy to rename it to something else, I just can't think of a better name

Reply:
My main "objection" is that I find the escaped path separators to be harder to read than the original patterns

<img width="811" alt="image" src="https://github.com/jestjs/jest/assets/1404810/58e952bf-c859-4e1e-bc5d-0900fd749095">

Is just more characters to ignore in order to read the pattern

Reply:
Oh ok so currently, jest renders the "no tests found" and the "ran tests matching" messages differently:
```console
$ jest foo/bar bar/baz --testPattern x/a

Pattern: foo/bar|bar/baz|x/a - 0 matches

Ran all test suites matching /foo\/bar|bar\/baz|x\/a/i.
```
1. Should these messages render the pattern the same way? Or should it be the case that the no-matches message shows a "pretty" version, and the tests-found message shows the actual regex?
2. If they should be rendered the same way, which way should they be rendered?

Reply:
I'd personally prefer them to render without escape characters - it's just easier to read (same as we change backslashes to forward slashes for windows in reporters and such). And consistency is always good! 🙂 

So 1. yes - 2. `Pattern`.

Reply:
Updated: d0d918dc10f95434c1e97b867bd631881b5c7d57